### PR TITLE
Fixes possible leak caused by concurrent add/remove of IMap interceptors

### PR DESCRIPTION
**Pull request by ahmetmircik**
Fixes the leak which is introduced in PR https://github.com/hazelcast/hazelcast/pull/3993

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        assert !(Thread.currentThread() instanceof PartitionOperationThread);
+
+        if (id2InterceptorMap.containsKey(id)) {
+            return;
+        }
+
+        Map<String, MapInterceptor> tmpMap = new HashMap<String, MapInterceptor>(id2InterceptorMap);
+        tmpMap.put(id, interceptor);
+
+        id2InterceptorMap = tmpMap;
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java]: Can you also wrap with `Collections.unmodifiableMap()`? Readers should not be able to modify.

```
id2InterceptorMap = Collections.unmodifiableMap(tmpMap);
```

File: hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        List<MapInterceptor> tmpInterceptors = new ArrayList<MapInterceptor>(interceptors);
+        tmpInterceptors.add(interceptor);
+
+        interceptors = tmpInterceptors;
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java]: Same here, you can wrap with `Collections.unmodifiableList()`?

Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java (hunk 1 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java]: done.

**Comment by mdogan** [comment]
:+1:

**Comment by mdogan** [comment]
Will you backport this?

**Comment by gurbuzali** [comment]
:+1:

**Comment by ahmetmircik** [comment]
@mdogan backport is on the way.

**Pull request merged**